### PR TITLE
Add BcolzArrayIterator

### PR DESCRIPTION
**Pull request by MPJansen**
Add an iterator for bcolzArrays to simplify streaming from pre-computed network outputs. This avoids having to load the entire array into ram.

Usage:
```
X = bcolz.open(path + 'train_convlayer_features.bc', mode='r')
y = bcolz.open(path + 'train_labels.bc', mode='r')
trn_batches = BcolzArrayIterator(X, y, batch_size=X.chunklen * batch_size, shuffle=True)
model.fit_generator(generator=trn_batches, samples_per_epoch=trn_batches.N, nb_epoch=1)
```
Thanks to R4mon: http://pastebin.com/y0DskiK6

### CODE REVIEW

**Comment by MPJansen** [comment]
Hi @racheltho,
took me a while to find the time, but I've rewritten the code a little.

I found out that Bcolz arrays are sliceable much like numpy arrays are, so I did away with the chunk reading in the original. This shouldnt hurt the maximum attainable performance that much, maybe a little when shuffling, Now though, shuffling actually gives you a random order of inputs instead of a random order of chunks (typically chunks are larger than the batch size so this would mean batches remain always the same)

Docs and tests added as per review,

**Comment by jph00** [comment]
Ah, well there's a good reason for all that chunks business, and that is that to grab a single slice, bcolz has to decompress the entire chunk. So the performance of randonly taking single slices is a disaster - unless you use a chunk size of 1 (in which case there's no need for this class - you can just use a basic numpy shuffle iterator).

I actually need to use this now, so I'll just make the necessary changes myself, if that's OK. BTW I found the original Kaggle forum thread where this was introduced, so I'll also add a credit to the original author for the code, and also to you for the docs/tests.

**Comment by MPJansen** [comment]
Correct, although I think that when shuffle is off, Bcolz caches the decompressed file so performance is ok.
The problem with the chunk reading is that shuffling isnt all that random, since the same features will end up together in a batch.

Feel free to make changes, I was unable to find the forum post after I closed the browser, I had only the link to pastebin ..

**Comment by MPJansen** [comment]
Ah I figured as much, although speed tests showed it was still faster than the original.

Is the fact that the same features always end up in the same batch not a problem? (this makes shuffling less random right)?

Please feel free to edit, would you happen to have a link to the original kaggle post? I seem to be unable to find it.

Looking forward to part 2 when it comes out for public viewing :)

**Comment by jph00** [comment]
There's not much else in the kaggle thread: https://www.kaggle.com/c/data-science-bowl-2017/discussion/27935

The issue with caching is that it's only going to really help here if the dataset if smaller than RAM. And if that's the case, you don't need to use this iterator at all, since you can just read in the whole array into memory!

So your performance test really needs to test a file that is, say, twice as big as RAM.

I'll close this PR and add the file directly to GH.

**Comment by MPJansen** [comment]
Hi @jph00, Will you be adding the iterator to the Part 1 repo as well?
If not, I'd reopen this and add a chunked version of the iterator (with correct StopIteration raise etc.)

**Comment by jph00** [comment]
I've been working on it all week and it keeps giving me wrong answers -
but only on large datasets. I just can't figure it out! When I use
shuffle=True and do fit_generator() on the full imagenet dataset, I get
totally wrong results. But when I repeat the exact same code on a subset
of 20,000 random images, I get great results. I don't want to commit it
until we have it working well...

**Comment by MPJansen** [comment]
That sounds frustrating,

If you'd like another set of eyes, I'd be happy to have a look, 
Does the same thing happen with the script that was doing the redundant copies? (the one from kaggle)

I think I noticed something similar during training on statefarm while using this (suddenly the labels didnt seem to match up with the features anymore) but I couldnt isolate it.

grts

**Comment by jph00** [comment]
I'd love help fixing it. I didn't make any changes beyond fixing up the
doctest a little and some minor refactoring. I just tried creating a
test that ran through 10,000 epochs and checked the labels were still
assigned correctly each epoch, and it passed. So I'm a little stuck...

**Comment by MPJansen** [comment]
So with the original script? Do you have a way of generating the wrong results too?

One thing I noticed is that the original iterator doesnt raise StopIteration when it completes a full run of the test samples, I do not really know what happens when you iterate out of bounds so to speak.
When I do:
`[v for v in bc_it]`
where bc_it is the iterator into some bcolz array, the iteration loop never stops (and fills up RAM really fast). 
Since you specify the amount of samples you train on in the generator, it is possible to ask more 

I'll start with some simple numpy arrays with known data and see as I do not have a server at home I can happily load/train Imagenet on ;)

**Comment by jph00** [comment]
That's actually the way that all the keras generators behave, I believe.
You just keep calling them and they keep looping through.

**Comment by MPJansen** [comment]
Right, I missed the reset line that zeros the batch index.
Although I did just notice that if you have no leftover_arrays, that is if the number of data entries is a multiple of the chunklen then the iterator returns an empty array every epoch. 
Because of this line:
`self.index_array = np.arange(self.X.nchunks + 1)`
Where the +1 is I think to deal with the trailing leftover_arrays.

I am unsure what happens if you return an empty array to keras though, because from what I can see the labels and weights would remain aligned with the data itself after the next iteration (when the iterator resets)

**Comment by jph00** [comment]
I tried running it on imagenet last night with something that recreated
the iterator from scratch after each epoch. That worked fine. Not sure
what that means exactly, I think it counts as progress!

**Comment by MPJansen** [comment]
Ha, great news.

I'll try and see what happens to Keras when you put in an empty array tonight. Could you check what the leftover_array size is on imagenet bcolz array?

**Comment by jph00** [comment]
OK I figured it out. It's a race condition - the concatenation and
return is not protected by the lock. The fix is to indent those lines.
I'll check it in to GH. Since I'm only using it for the course in part
2, do you mind if I commit it there only? Otherwise I'm worried that
it'll confuse students to see a class that's never used... (I can give
you access to the part 2 GH repo if you want it).

**Comment by MPJansen** [comment]
Hey @jph00 

That is great! Ah now I see, I thought that did not matter as it did not interact with the bcolz array directly.
I need to study up on concurrency in Python ...
Access would be terrific :) 
I'll spend some time this weekend to eliminate the concatenation (because as @racheltho pointed out, it really takes quite a long time to copy all those values twice) and see if I can get rid of the behaviour that returns an empty array at the end of an epoch.

**Comment by jph00** [comment]
OK will send you details - what's your email or twitter (if twitter,
please follow @jeremyphoward so I can DM)?

**Comment by MPJansen** [comment]
I've joined the twitterverse seconds ago,
have followed you under the alias MatthijsPJ!

Thanks again for all the hard work you guys put into this :)
